src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -836,8 +836,9 @@ public void recover(RecoveryHandler recoveryHandler) throws EngineException {
             maybeFailEngine("recovery phase 3", e);
             throw new RecoveryEngineException(shardId, 3, "Execution failed", wrapIfClosed(e));
         } finally {
-            Releasables.close(success, onGoingRecoveries, writeLock, phase1Snapshot,
-                    phase2Snapshot, phase3Snapshot); // hmm why can't we use try-with here?
+            // close the snapshots first to release the reference to the translog file, so a flush post recovery can delete it
+            Releasables.close(success, phase1Snapshot, phase2Snapshot, phase3Snapshot,
+                    onGoingRecoveries, writeLock); // hmm why can't we use try-with here?
```

Comment:
The problem with try with resources clause is that we release the phase1 and phase2 snapshots which are defined earlier in the scope.  I can make the write lock use a try-with but this requires another catch to convert the exception the close method throws into a RecoveryEngineException. I'm not sure it's worth it.